File: reply_service.py
from datetime import datetime
from models import Handle, Reply, DenyWord
from app import db

class ReplyService:
    @staticmethod
    def store_reply(handle: str, reply_id: str, content: str) -> Reply:
        """Store a new reply and check if it should be hidden."""
        # Get or create handle
        handle_obj = Handle.query.filter_by(handle=handle).first()
        if not handle_obj:
            handle_obj = Handle(handle=handle)
            db.session.add(handle_obj)
            db.session.flush()
        
        # Create reply
        reply = Reply(
            reply_id=reply_id,
            content=content,
            handle=handle_obj
        )
        
        # Check against existing denywords
        denywords = DenyWord.query.filter_by(handle_id=handle_obj.id).all()
        for denyword in denywords:
            if denyword.word.lower() in content.lower():
                reply.is_hidden = True
                reply.hidden_at = datetime.utcnow()
                reply.hidden_by_word = denyword.word
                break
        
        db.session.add(reply)
        db.session.commit()
        return reply

    @staticmethod
    def add_denyword(handle: str, word: str) -> tuple[DenyWord, list[Reply]]:
        """Add a new denyword and process existing replies."""
        # Get or create handle
        handle_obj = Handle.query.filter_by(handle=handle).first()
        if not handle_obj:
            handle_obj = Handle(handle=handle)
            db.session.add(handle_obj)
            db.session.flush()
        
        # Create denyword
        denyword = DenyWord(word=word, handle=handle_obj)
        db.session.add(denyword)
        
        # Find and update existing replies that match
        newly_hidden = []
        replies = Reply.query.filter_by(
            handle_id=handle_obj.id,
            is_hidden=False
        ).all()
        
        for reply in replies:
            if word.lower() in reply.content.lower():
                reply.is_hidden = True
                reply.hidden_at = datetime.utcnow()
                reply.hidden_by_word = word
                newly_hidden.append(reply)
        
        db.session.commit()
        return denyword, newly_hidden

    @staticmethod
    def get_hidden_replies(handle: str) -> list[Reply]:
        """Get all hidden replies for a handle."""
        handle_obj = Handle.query.filter_by(handle=handle).first()
        if not handle_obj:
            return []
        return Reply.query.filter_by(
            handle_id=handle_obj.id,
            is_hidden=True
        ).order_by(Reply.hidden_at.desc()).all()

File: services/replyService.js
const { Handle, Reply, DenyWord } = require('../models');
const winston = require('winston');

const logger = winston.createLogger({
    level: 'info',
    format: winston.format.json(),
    transports: [new winston.transports.Console()]
});

class ReplyService {
    static async storeReply(handle, replyId, content) {
        try {
            // Get or create handle
            const [handleObj] = await Handle.findOrCreate({
                where: { handle }
            });

            // Check against existing denywords
            const denywords = await DenyWord.findAll({
                where: { HandleId: handleObj.id }
            });

            const reply = await Reply.create({
                replyId,
                content,
                HandleId: handleObj.id
            });

            // Check content against denywords
            for (const denyword of denywords) {
                if (content.toLowerCase().includes(denyword.word.toLowerCase())) {
                    reply.isHidden = true;
                    reply.hiddenAt = new Date();
                    reply.hiddenByWord = denyword.word;
                    await reply.save();
                    break;
                }
            }

            return reply;
        } catch (error) {
            logger.error('Error storing reply:', error);
            throw error;
        }
    }

    static async addDenyword(handle, word) {
        try {
            // Get or create handle
            const [handleObj] = await Handle.findOrCreate({
                where: { handle }
            });

            // Create denyword
            const denyword = await DenyWord.create({
                word,
                HandleId: handleObj.id
            });

            // Find and update existing replies that match
            const replies = await Reply.findAll({
                where: {
                    HandleId: handleObj.id,
                    isHidden: false
                }
            });

            const newlyHidden = [];
            for (const reply of replies) {
                if (reply.content.toLowerCase().includes(word.toLowerCase())) {
                    reply.isHidden = true;
                    reply.hiddenAt = new Date();
                    reply.hiddenByWord = word;
                    await reply.save();
                    newlyHidden.push(reply);
                }
            }

            return { denyword, newlyHidden };
        } catch (error) {
            logger.error('Error adding denyword:', error);
            throw error;
        }
    }

    static async getHiddenReplies(handle) {
        try {
            const handleObj = await Handle.findOne({
                where: { handle }
            });

            if (!handleObj) {
                return [];
            }

            return Reply.findAll({
                where: {
                    HandleId: handleObj.id,
                    isHidden: true
                },
                order: [['hiddenAt', 'DESC']]
            });
        } catch (error) {
            logger.error('Error getting hidden replies:', error);
            throw error;
        }
    }
}

module.exports = ReplyService;
